lib/SILOptimizer/Transforms/ArrayElementValuePropagation.cpp

```diff
@@ -154,31 +181,67 @@ bool ArrayAllocation::mapInitializationStores() {
 /// allocated.
 bool ArrayAllocation::isInitializationWithKnownElements() {
   ArraySemanticsCall Uninitialized(Alloc, "array.uninitialized");
-  if (Uninitialized &&
-      (ArrayValue = Uninitialized.getArrayValue()) &&
-      (ElementBuffer = Uninitialized.getArrayElementStoragePointer()))
+  if (Uninitialized && (ArrayValue = Uninitialized.getArrayValue()) &&
+      (ElementBuffer = Uninitialized.getArrayElementStoragePointer())) {
+    Substitutions =
+        ArrayValue->getType().gatherAllSubstitutions(Alloc->getModule());
     return mapInitializationStores();
+  }
 
   return false;
 }
 
+bool ArrayAllocation::replacementsAreValid() {
+  auto ElementCount = ElementValueMap.size();
+  llvm::SmallVector<SILValue, 4> ElementValueVector;
+
+  if (ElementCount > APPEND_CONTENTSOF_REPLACEMENT_VALUES_MAX)
+    return false;
+
+  for (unsigned i = 0; i < ElementCount; ++i) {
+    // This means that elements aren't contiguous
+    if (!ElementValueMap.count(i))
+      return false;
+  }
+
+  return true;
+}
+
 /// Propagate the elements of an array literal to get_element method calls on
 /// the same array.
 ///
 /// We have to prove that the array value is not changed in between the
 /// creation and the method call to get_element.
-bool ArrayAllocation::findValueReplacements() {
+bool ArrayAllocation::findReplacements() {
   if (!isInitializationWithKnownElements())
     return false;
 
   // The array value was stored or has escaped.
   if (!analyzeArrayValueUses())
     return false;
 
-  // No count users.
-  if (GetElementCalls.empty())
+  if (GetElementCalls.empty() && AppendContentsOfCalls.empty())
```

Comment:
Please be consistent about using braces or not using braces with if statements on single lines. The rest of this looks good! Maybe add some comments describing how you are "peeling" off possible conditions causing your final bit of work to be valid?